### PR TITLE
Add special Comparer/EqualityComparer for Nullable<Enum>

### LINKED ISSUE #67842: EqualityComparer<TEnum?>.Default allocates
### Description

`EqualityComparer<TEnum?>.Default` (when `T` is a *nullable enum*) is resolved to `ObjectEqualityComparer<TEnum?>`, which boxes the enum when performing equality comparison. This causes an unnecessary allocation.

Here is a reproduction:

```C#
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace ConsoleApp;

public static class Program
{
    public static void Main()
    {
        _ = EqualityComparer<DayOfWeek?>.Default.Equals(DayOfWeek.Friday, DayOfWeek.Friday); // Warm-up

        var before = GC.GetAllocatedBytesForCurrentThread();
        _ = EqualityComparer<DayOfWeek?>.Default.Equals(DayOfWeek.Friday, DayOfWeek.Friday);
        var after = GC.GetAllocatedBytesForCurrentThread();

        Console.WriteLine(RuntimeInformation.FrameworkDescription);
        Console.WriteLine(after > before ? "Allocates" : "Allocation-free");
    }
}
```

This results in:

```
.NET 7.0.0-preview.2.22152.2
Allocates
```

If you replace `DayOfWeek?` with `DayOfWeek`, or if you use a type such as `int?` instead, the code will print `Allocation-free`.

### Configuration

I reproduced this on .NET 6 and on .NET 7 preview 2 on Windows x64, but this should be independent of the OS and architecture.

### Regression?

This is not a regression. It also reproduces on .NET Framework 4.8.

### Analysis

This is roughly due to the fact that [`ComparerHelpers.CreateDefaultEqualityComparer`](https://github.com/dotnet/runtime/blob/7bfc61b970e28f94782ef7c0cbcbbbc94ef9f5eb/src/coreclr/System.Private.CoreLib/src/System/Collections/Generic/ComparerHelpers.cs#L116) special-cases both `Nullable<T>` *and* enums, but does not special-case nullable enums. This code path ends up using `ObjectEqualityComparer<T>`.

I'm willing to provide a PR to fix this if you'd like.

### PR DESCRIPTION
**Pull request by EgorBo**
Fixes https://github.com/dotnet/runtime/issues/67842

Example:
```csharp
bool Test(DayOfWeek? d1, DayOfWeek? d2) => EqualityComparer<DayOfWeek?>.Default.Equals(d1, d2);
```
Codegen diff: https://www.diffchecker.com/7qPt1KtF
It doesn't box (allocate) any more. The final codegen is still a bit suboptimal but it's a different problem (`Nullable<>` is not promoted)

Same codegen for e.g.:
```csharp
record struct Test(DayOfWeek? Day);

bool Test(Test t1, Test t2) => t1.Equals(t2);
```

### CODE REVIEW

**Comment by jkotas** [comment]
Would it be better to instead figure out how to optimize the boxing from the constrained calls to Enum.Equals?

File: src/coreclr/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.CoreCLR.cs [hunk 1]
```diff
@@ -183,9 +183,34 @@ internal override int LastIndexOf(T[] array, T value, int startIndex, int count)
...
             {
-                if (System.Runtime.CompilerServices.RuntimeHelpers.EnumEquals(array[i], value)) return i;
+                if (RuntimeHelpers.EnumEquals(array[i], value)) return i;
             }
             return -1;
         }
     }
+
+    // Instantiated internally by VM
+    internal sealed class NullableEnumEqualityComparer<T> : EqualityComparer<T?> where T : struct, Enum
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.CoreCLR.cs:193]: This needs tests and extra code to ensure that the binary serialization continues to work.

**Comment by jkotas** [comment]
This also needs to be handled in src\coreclr\tools\Common\TypeSystem\IL\Stubs\ComparerIntrinsics.cs

**Comment by jkotas** [comment]
Do we want to add it for Mono too? We generally try to add optimizations like these for all runtimes.

**Comment by jkotas** [comment]
> Let's intrinsify Nullable<>.Equals instead

How is this fixing the problem with boxing in EqualityComparer from the linker issue?

**Comment by EgorBo** [comment]
> > Let's intrinsify Nullable<>.Equals instead
> 
> How is this fixing the problem with boxing in EqualityComparer from the linker issue?

@jkotas it's called by the ObjectEqualityComparer.

The current problem that
```
bool Test(DayOfWeek? d1, DayOfWeek? d2) => EqualityComparer<DayOfWeek?>.Default.Equals(d1, d2);
```
ends up as `Nullable<>.Equals(object)` call which is then inlined into `Enum.Equals(object)` where input enum becomes a multi-use boxing we're not able to handle currently https://github.com/dotnet/runtime/issues/9118

so while this PR indeed works it's quite verbose and doesn't handle CompareTo so I think it's better to rely on a custom equalitycomparer for now

**Comment by jkotas** [comment]
> EqualityComparer<DayOfWeek?>.Default.Equals(d1, d2);
> ends up as Nullable<>.Equals(object) call which is then inlined into Enum.Equals(object)

It should end up as `NullableEqualityComparer.Equals`

`NullableEqualityComparer.Equals` calls `Enum.Equals` here: https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs#L107 . 

I do not see anything calling `Nullable.Equals` on this path.

**Comment by EgorBo** [comment]
> It should end up as `NullableEqualityComparer.Equals`

it does not currently because it doesn't pass this check https://github.com/dotnet/runtime/blob/7bfc61b970e28f94782ef7c0cbcbbbc94ef9f5eb/src/coreclr/vm/jitinterface.cpp#L8822-L8827 🤔

**Comment by jkotas** [comment]
Ok, make sense.

**Comment by jkotas** [comment]
I am wondering whether we can just fix the existing `NullableEqualityComparer` instead of introducing a new one. Drop `IEquatable<T>` constraint from `NullableEqualityComparer<T>` and change `x.value.Equals(y.value);` in `NullableEqualityComparer` implementation to `EqualityComparer<T>.Default.Equals(x, y)`?

**Comment by EgorBo** [comment]
> I am wondering whether we can just fix the existing `NullableEqualityComparer` instead of introducing a new one. Drop `IEquatable<T>` constraint from `NullableEqualityComparer<T>` and change `x.value.Equals(y.value);` in `NullableEqualityComparer` implementation to `EqualityComparer<T>.Default.Equals(x, y)`?

Great idea! Let's see how it goes...

File: src/coreclr/vm/jitinterface.cpp [hunk 2]
```diff
@@ -8931,12 +8931,8 @@ CORINFO_CLASS_HANDLE CEEInfo::getDefaultEqualityComparerClassHelper(CORINFO_CLAS
     if (Nullable::IsNullableType(elemTypeHnd))
     {
         Instantiation nullableInst = elemTypeHnd.AsMethodTable()->GetInstantiation();
-        TypeHandle iequatable = TypeHandle(CoreLibBinder::GetClass(CLASS__IEQUATABLEGENERIC)).Instantiate(nullableInst);
-        if (nullableInst[0].CanCastTo(iequatable))
-        {
-            TypeHandle resultTh = ((TypeHandle)CoreLibBinder::GetClass(CLASS__NULLABLE_EQUALITYCOMPARER)).Instantiate(nullableInst);
-            return CORINFO_CLASS_HANDLE(resultTh.GetMethodTable());
-        }
+        TypeHandle resultTh = ((TypeHandle)CoreLibBinder::GetClass(CLASS__NULLABLE_EQUALITYCOMPARER)).Instantiate(nullableInst);
```
Comment (jkotas) [inline src/coreclr/vm/jitinterface.cpp:8934]: We also need a matching change in src\coreclr\tools\Common\TypeSystem\IL\Stubs\ComparerIntrinsics.cs

Comment (jkotas) [reply src/coreclr/vm/jitinterface.cpp:8934]: GetComparerForType

**Review by jkotas — APPROVED** [review]
Could you please check asm diffs for the existing NullableEqualityComparer uses to make sure there is no significant code quality regression?

Otherwise looks good.

**Comment by EgorBo** [comment]
@jkotas Thanks! Sorry for the delay between changes - I was on a month-long "vacation".

The codegen diff is more or less the same:
1) Nullable Primitives - no diffs
2) Nullable Struct implementing IEquatable/IComparable - a few redundant movs
3) Nullable structs and enums - much better codegen and no allocations (boxing)

Had to remove a test in Libraries - it used to rely on fact that a round-trip binary serialization should produce the same or assignable object but it doesn't work now.

e.g.

```csharp
EqualityComparer<T> orig = EqualityComparer<T>.Default;
bf.Serialize(s, orig);
s.Position = 0;
object result = bf.Deserialize(s);
Assert.IsAssignableFrom<EqualityComparer<T>>(result);
```
For `T` being `Nullable<MyInt16Enum>` we used to get `ObjectEqualityComparer<Nullable<MyInt16Enum>>` object from .Default but now we get `NullableEqualityCompdarer<MyInt16Enum>`
which after a round-trip serialization becomes `ObjectEqualityComparer<Nullable<MyInt16Enum>>` due to backward compatibility hack

**Comment by EgorBo** [comment]
Failure is unrelated (https://github.com/dotnet/runtime/issues/68511)

**Pull request merged**